File: app_java/src/main/java/vlab/android/architecture/MyApplication.java
package vlab.android.architecture;

import com.squareup.leakcanary.LeakCanary;

import vlab.android.architecture.di.AppModule;
import vlab.android.architecture.di.DaggerAppComponent;
import vlab.android.architecture.di.module.NetworkModule;
import vlab.android.common.DaggerCommonApplication;
import vlab.android.common.util.LogUtils;

/**
 * Created by Vinh Tran on 2/11/18.
 */
public class MyApplication extends DaggerCommonApplication {
    private static MyApplication sInstance = null;

    public static MyApplication getInstance() {
        return sInstance;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        LogUtils.d(getClass().getSimpleName(), ">>> MyApplication onCreate");
        if (LeakCanary.isInAnalyzerProcess(this)) {
            // This process is dedicated to LeakCanary for heap analysis.
            // You should not init your app in this process.
            return;
        }
        LeakCanary.install(this);
        sInstance = this;

        // init log
        LogUtils.init(vlab.android.common.BuildConfig.DEBUG);
    }

    @Override
    protected void buildDaggerAppComponent() {
        DaggerAppComponent.builder()
                .appModule(new AppModule(this))
                .networkModule(new NetworkModule("https://api.github.com/")) // github
                .build()
                .inject(this);
    }
}

File: app_kotlin/src/main/java/vlab/android/architecture/MyApplication.kt
package vlab.android.architecture

import android.app.Application
import org.koin.android.ext.koin.androidContext
import org.koin.android.ext.koin.androidLogger
import org.koin.core.context.startKoin
import vlab.android.architecture.di.koin.apiModule
import vlab.android.architecture.di.koin.networkModule
import vlab.android.architecture.di.koin.repositoryModule
import vlab.android.architecture.di.koin.viewModelModule
import vlab.android.common.util.LogUtils

/**
 * Created by Vinh Tran on 2/11/18.
 */
class MyApplication : Application() {

    companion object {
        private lateinit var sInstance: MyApplication

        fun getInstance() : MyApplication {
            return sInstance
        }
    }

    override fun onCreate() {
        super.onCreate()
        sInstance = this
        LogUtils.d(javaClass.simpleName, ">>> MyApplication onCreate")
        // init log
        LogUtils.init(BuildConfig.DEBUG)

        // start Koin context
        startKoin {
            androidContext(this@MyApplication)
            androidLogger()
            modules(
                networkModule,
                apiModule,
                repositoryModule,
                viewModelModule
            )
        }
    }

}